mem/buffer_slice.go

```diff
@@ -219,8 +224,56 @@ func (w *writer) Write(p []byte) (n int, err error) {
 
 // NewWriter wraps the given BufferSlice and BufferPool to implement the
 // io.Writer interface. Every call to Write copies the contents of the given
-// buffer into a new Buffer pulled from the given pool and the Buffer is added to
-// the given BufferSlice.
+// buffer into a new Buffer pulled from the given pool and the Buffer is
+// added to the given BufferSlice.
 func NewWriter(buffers *BufferSlice, pool BufferPool) io.Writer {
 	return &writer{buffers: buffers, pool: pool}
 }
+
+// ReadAll reads from r until an error or EOF and returns the data it read.
+// A successful call returns err == nil, not err == EOF. Because ReadAll is
+// defined to read from src until EOF, it does not treat an EOF from Read
+// as an error to be reported.
+// A failed call returns a non-nil error and could return partially read
+// buffers. It is the responsibility of the caller to free this buffer.
```

Comment:
This is surprising behavior.  The one saving grace is that freeing buffers is optional -- GC will take them away if you forget.  If not for that, I would say this is definitely not OK.

I highly doubt gRPC would ever want the partially data, and I'm curious why you want it, too.

Reply:
This behavior matches [`io.ReadAll()`](https://pkg.go.dev/io#ReadAll). Reasons to do it this way:

- A drop-in replacement for `io.ReadAll()` (behavior-wise).
- Sometimes you may need the read data regardless if there was an error or not.

Example of the last point: a proxy forwarding a response from an upstream server. It must send everything it got from the upstream and then return an error (return == maybe RST the connection or something else, protocol-specific). Isn't e.g. gRPC streaming response client the same? It reads and provides the client with messages even if it already got an error after the message. This is also similar to how HTTP/1.0 or 1.1 `Connection: close` works. Sometimes you don't want `Transfer-Encoding: chunked` and prefer the underlying connection to be closed on EOF instead of chunking.

Reply:
Are you using this completely outside of gRPC?  This package isn't really intended as a generic thing, it's intended to meet our needs for our use cases.

Reply:
@dfawley  My project uses gRPC "a lot". Most of the uses of this function are not related to gRPC directly at the moment but this may change.

In my opinion more flexible behavior is a better choice since it's not an `internal` package/function. gRPC might need it later but it'd be impossible to change - you'll have to introduce a new function.

Having said the above, I have no new things to say =) I'm happy to change it to free the buffers on error. My ultimate goal here is to get rid of the unnecessary allocations inside of gRPC. I can absolutely live with a copy of this function in my codebase and use the modified version for my own purposes.

Please let me know how you want to proceed.

Reply:
OK, this is fine, but I have some minor edits I'll make to the docstring to make it stand out a little more.